tools/trimmomatic.py

```diff
@@ -15,6 +20,58 @@ def __init__(self, install_methods=None):
             install_methods.append(tools.CondaPackage(TOOL_NAME, version=TOOL_VERSION))
         tools.Tool.__init__(self, install_methods=install_methods)
 
-    def execute(self, *args):
-        cmd = [self.install_and_get_path()] + args
-        subprocess.check_call(cmd)
+    def execute(self,
+        inFastq1,
+        inFastq2,
+        pairedOutFastq1,
+        pairedOutFastq2,
+        clipFasta,
+        unpairedOutFastq1=None,
+        unpairedOutFastq2=None,
+        leading_q_cutoff=15,
+        trailing_q_cutoff=15,
+        minlength_to_keep=30,
+        sliding_window_size=4,
+        sliding_window_q_cutoff=25
+    ):
+        '''Trim read sequences with Trimmomatic.'''
+        trimmomaticPath = self.install_and_get_path()
+        unpairedFastq1 = unpairedOutFastq1 or util.file.mkstempfname()
+        unpairedFastq2 = unpairedOutFastq2 or util.file.mkstempfname()
+        javaCmd = [trimmomaticPath]
+
+        if inFastq2 is None or os.path.getsize(inFastq2) < 10:
+            # Unpaired reads
+            javaCmd.extend([
+                    'SE', '-phred33',
```

Comment:
I think I prefer autodetection where possible, partly because this is really a fastq-specific problem, and the user's input files (BAM) can always be paired, single, or a mixture of both. If it's a mixture, I wouldn't know how to expose the parameter to the user.